test/Middleware/Authentication/AppInstallationTokenProviderTest.cs

```diff
@@ -0,0 +1,117 @@
+{
+    private readonly Mock<IGitHubAppTokenProvider> _mockGitHubAppTokenProvider;
+    private readonly RSA _privateKey;
+    private readonly AppInstallationTokenProvider _tokenProvider;
+
+    public AppInstallationTokenProviderTests()
+    {
+        _mockGitHubAppTokenProvider = new Mock<IGitHubAppTokenProvider>();
+        _privateKey = RSA.Create();
+        _tokenProvider = new AppInstallationTokenProvider("clientId", _privateKey, "installationId", _mockGitHubAppTokenProvider.Object);
+    }
+
+    [Fact]
+    public void Constructor_InitializesTokenProviderWithClientId()
+    {
+        var privateKey = RSA.Create();
+        var tokenProvider = new AppInstallationTokenProvider("clientId", privateKey, "installationId", new GitHubAppTokenProvider());
+
+        Assert.NotNull(tokenProvider);
+    }
+
+    [Fact]
+    public void Constructor_InitializesTokenProviderWithAppId()
+    {
+        var privateKey = RSA.Create();
+        var tokenProvider = new AppInstallationTokenProvider(1, privateKey, "installationId", new GitHubAppTokenProvider());
+
+        Assert.NotNull(tokenProvider);
+    }
+
+    [Fact]
+    public async Task GetAuthorizationTokenAsync_TokenIsEmpty_FetchesNewToken()
+    {
+        var requestUri = new Uri("https://api.github.com");
+        _mockGitHubAppTokenProvider.Setup(x => x.CreateTokenDescriptor(_privateKey, "clientId", It.IsAny<DateTime>())).Returns(new SecurityTokenDescriptor());
+        _mockGitHubAppTokenProvider.Setup(x => x.CreateJsonWebToken(It.IsAny<SecurityTokenDescriptor>())).Returns("mockJwt");
+        _mockGitHubAppTokenProvider.Setup(x => x.GetGitHubAccessTokenAsync(It.IsAny<string>(), "mockJwt", "installationId")).ReturnsAsync("newAccessToken");
+
+        var token = await _tokenProvider.GetAuthorizationTokenAsync(requestUri);
+
+        Assert.Equal("newAccessToken", token);
+        _mockGitHubAppTokenProvider.Verify(x => x.CreateTokenDescriptor(_privateKey, "clientId", It.IsAny<DateTime>()), Times.Once);
+        _mockGitHubAppTokenProvider.Verify(x => x.CreateJsonWebToken(It.IsAny<SecurityTokenDescriptor>()), Times.Once);
+        _mockGitHubAppTokenProvider.Verify(x => x.GetGitHubAccessTokenAsync(It.IsAny<string>(), "mockJwt", "installationId"), Times.Once);
+    }
+
+    // TODO(kfcampbell): i don't love that the below tests rely on reflection; perhaps there's a better
+    // way to structure the code to avoid it? perhaps we could set private fields as internal?
+    [Fact]
+    public async Task GetAuthorizationTokenAsync_TokenIsExpired_FetchesNewToken()
+    {
+        var requestUri = new Uri("https://api.github.com");
+        var expiredTokenDescriptor = new SecurityTokenDescriptor { Expires = DateTime.UtcNow.AddMinutes(-2) };
+        var tokenDescriptorField = _tokenProvider.GetType()
+            .GetField("_tokenDescriptor", System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance);
+
+        if (tokenDescriptorField != null)
```

Comment:
I mentioned this in the comment above, but I really don't like the reflection here. Please let me know if you have better ideas for testability (using `internal` instead of `private` maybe?).

Reply:
I think internal should be fine here.. where you thinking about using a decorator to make it available to the test project/file.. i.e.

`[assembly: InternalsVisibleTo("test")]` or something - I'm good with either way.  While not inherently "bad", I'd like to stay away from reflection if possible.

Reply:
Agreed on staying away from reflection. I'm not sure of the exact incantation required to make the internals testable; if a decorator is the way to go, that's definitely better than reflection. Perhaps we can pair on this tomorrow?